Plugins/Public/base_plugin/FactoryModule.cpp

```diff
@@ -1,81 +1,119 @@
+			uint quantity = i.second;
 
-			for (map<uint, uint>::iterator i = active_recipe.consumed_items.begin();
-				i != active_recipe.consumed_items.end(); ++i)
+			const GoodInfo* gi = GoodList::find_by_id(good);
+			if (gi)
 			{
-				uint good = i->first;
-				uint quantity = i->second;
+				info += openLine + L"- " + stows(itos(quantity)) + L"x " + HkGetWStringFromIDS(gi->iIDSName);
+				if (quantity > 0 && base->HasMarketItem(good) < active_recipe.cooking_rate)
+				{
+					info += L" [Out of stock]";
+				}
+			}
+		}
+		if (active_recipe.credit_cost)
+		{
+			info += openLine + L" - Credits $" + UIntToPrettyStr(active_recipe.credit_cost);
+			if (base->money < active_recipe.credit_cost)
+			{
+				info += L" [Insufficient cash]";
+			}
+		}
+		if (!sufficientCatalysts)
+		{
+			info += openLine + L"Insufficient catalysts/workforce";
+		}
+		if (!active_recipe.catalyst_items.empty())
+		{
+			info += openLine + L"Needed catalysts:";
+			for (const auto& catalyst : active_recipe.catalyst_items)
+			{
+				uint good = catalyst.first;
+				uint quantity = catalyst.second;
 
-				const GoodInfo *gi = GoodList::find_by_id(good);
+				const GoodInfo* gi = GoodList::find_by_id(good);
 				if (gi)
 				{
-					info += L"<PARA/><TEXT>      - " + stows(itos(quantity)) + L"x " + HkGetWStringFromIDS(gi->iIDSName);
-					if (quantity > 0 && base->HasMarketItem(good) < active_recipe.cooking_rate)
-						info += L" [Out of stock]";
-					info += L"</TEXT>";
+					info += openLine + L" - " + stows(itos(quantity)) + L"x " + HkGetWStringFromIDS(gi->iIDSName);
+					uint presentAmount = base->HasMarketItem(good);
+					if (presentAmount < quantity)
+					{
+						info += L" [Need " + stows(itos(quantity - presentAmount)) + L" more]";
+					}
 				}
 			}
 		}
-		info += L"<TEXT>";
-	}
-	else
-	{
-		info += FACTORY_NAMES[type];
-		info += L" - Pending " + stows(itos(build_queue.size())) + L" items ";
-		if (active_recipe.nickname)
+		if (!active_recipe.catalyst_workforce.empty())
 		{
-			info = L" - Building " + Status + active_recipe.infotext + L". Waiting for:";
-
-			for (map<uint, uint>::iterator i = active_recipe.consumed_items.begin();
-				i != active_recipe.consumed_items.end(); ++i)
+			info += openLine + L"Needed workforce:";
+			for (const auto& worker : active_recipe.catalyst_workforce)
 			{
-				uint good = i->first;
-				uint quantity = i->second;
+				uint good = worker.first;
+				uint quantity = worker.second;
 
-				const GoodInfo *gi = GoodList::find_by_id(good);
+				const GoodInfo* gi = GoodList::find_by_id(good);
 				if (gi)
 				{
-					info += L" " + stows(itos(quantity)) + L"x " + HkGetWStringFromIDS(gi->iIDSName);
+					info += openLine + L" - " + stows(itos(quantity)) + L"x " + HkGetWStringFromIDS(gi->iIDSName);
+					uint presentAmount = base->HasMarketItem(good);
+					if (presentAmount < quantity)
+					{
+						info += L" [Need " + stows(itos(quantity - presentAmount)) + L" more]";
+					}
 				}
 			}
 		}
```

Comment:
doesn't the return below here need to add a final `</TEXT>` if in XML mode?

Reply:
See above